Add UI to restore previous versions of a tracing

daniel-wer opened the pull request:
This PR adds UI to restore older versions of a skeleton tracing. Restoring an older version simply creates a "new version" in the version history of the tracing - so it is possible to revert the reversion :)
The version preview and restore reuses functionality that we implemented for the task hot swap.

![version-restore-highlight](https://user-images.githubusercontent.com/1702075/45428378-6842d380-b6a1-11e8-88c2-e4ffcd762cd5.png)


### ToDo:
- [x] @philippotto In the version restore view, when scrolling, I would like to keep the headline on top and only scroll the version entries. However, I'm not css/html-savy enough to accomplish this. Any ideas?
- [x] @philippotto I'm having trouble with the flow types for the updateActions that are sent by the server. The server sends the same updateActions that we are sending, BUT there's an additional `actionTimestamp` (and maybe `version`). I tried to use flows $Call syntax to adapt our types, but the resulting flow type looks wrong to me (when looking at it through the type-at-pos functionality). I also can't get rid of the remaining flow errors.
- [x] @fm3 The versions are acquired through the datastore `updateActionLog` route, which returns batches of update actions. However, those batches are missing a version number. Currently the frontend is computing those versions, assuming the last entry is version 1, and counting up. Ideally, the backend would send the version number for each batch.
- [ ] We need to decide whether we want to group subsequent `createNode` actions by time interval or so, to avoid having too many of those `createNode` actions spamming the version restore view.

### URL of deployed dev instance (used for testing):
- https://restoreoldversions.webknossos.xyz

### Steps to test:
- Test the version restore mode
  - Create a skeleton tracing and trace a bit, split/merge a tree, delete a tree, whatever.
  - Then open the version restore view by clicking `Restore older version` in the dropdown next to the save button. The tracing should be saved and in read only mode now.
  - Preview any version by selecting it. You should not be able to modify the tracing.
  - Now there are two options:
    - Either close the version restore view. The newest version should be loaded again and you should be able to modify it and save.
    - Or select `Restore this version` for an older version. The old version should be loaded and you should be able to modify it and save. Opening the version restore view again should show that you reverted to a previous version.
- Create a volume tracing - the option to open the version restore view should not be visible.

### Issues:
- fixes #51 :open_mouth: 

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [ ] ~Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable~
- [x] Updated [documentation](../blob/master/docs) if applicable
- [x] Needs datastore update after deployment
- [x] Ready for review

daniel-wer commented on app/assets/javascripts/oxalis/model/sagas/update_actions.js:
```diff
@@ -135,12 +133,40 @@ export type UpdateAction =
+};
+type AsServerAction<A> = $Call<AddServerValuesFn, A>;
+
+// Since flow does not provide ways to perform type transformations on the
```
@philippotto Thanks!!! :pray:

philippotto commented on app/assets/javascripts/oxalis/view/version_view.js:
```diff
@@ -17,14 +16,25 @@ import { setAnnotationAllowUpdateAction } from "oxalis/model/actions/annotation_
+  T,
+) => {
+  name: $PropertyType<T, "name">,
+  value: { ...$PropertyType<T, "value">, actionTimestamp: number, version: number },
```
In general, this already looks very nice. Unfortunately, flow cannot understand that `$PropertyType<T, "name">` and `$PropertyType<T, "value">` should always refer to the same Union-part. That's why it thinks that name can be "updateTracing or deleteNode or ..." while simultaneously thinking "value can be X or Y or Z". It doesn't derive the constraint "if name is updateTracing, value is X".

I didn't find a nice way to express what we want, which is why I ended up defining a `ServerUpdateAction` which is a union of `AsServerType<UpdateTracing`-style parts. Additionally, it was necessary to add extra type-checks, since the typings of `_.groupBy` are not expressive enough (not sure whether they can be improved or if flow is too limited here, too).

philippotto commented:
> However, I broke it immediately by opening a new tracing, creating five nodes (hitting save after each creation) and opening the view :(

My bad, that's due to the changes I committed. Fixed it now :)

philippotto commented:
> However, I broke it immediately by opening a new tracing, creating five nodes (hitting save after each creation) and opening the view :(

My bad, that's due to the changes I committed. Fixed it now :)

daniel-wer commented:
What :joy: I reject all blame :grin: In all seriousness, I think you broke it with your flow fix commit.

daniel-wer commented:
Thanks for fixing! :)

daniel-wer commented:
@philippotto Thanks a lot for your help with the flow types and flex boxes :)

> - If the version-view is open, the tracing becomes read-only. This is not super obvious in my opinion. I added a hint to the text below the version-history-headline, but I'm not sure whether this enough. Do you think we can/should we make it more prominent?
> - Does the "copy to my account" button work when being in read-only mode?

It did not work as expected, same for the Download and Merge functionality, which is why I hide all of the tracing actions now. I added an Alert in its place and made the description in the version restore view an Alert as well. Not too intrusive but noticeable in my opinion :)

![version-restore-info](https://user-images.githubusercontent.com/1702075/45427726-e2725880-b69f-11e8-99a6-d4f9afe6f999.png)

> When reverting to a specific version, the version entry says "Reverted to version X". However, the version numbers are mentioned nowhere else in the view. Ideally, each entry should have the version number shown somewhere. But maybe you wanted to add this after the back-end provides this data?

I added the version number :)

> If I create a node, the corresponding version says "createNode and two other entries". What are the two other entries? update tracing due to moving? Also, we might want to map "createNode" to a non-camel-case variant.

I added descriptions for the other cases as well (createNode, updateTreeGroups, ...). The only thing that's "problematic" is updateTracing, as that could mean the user switched positions, or updated a comment, or updated a branchpoint, or ... I'd say a generic description is fine for now and we can always improve later.

I also added a scroll-into-view for the version restore view as I noticed that it was hidden to the right side if the screen was not big enough (because the tracing views were scaled too big).

philippotto commented:
Awesome! These changes really top it off :)

> - [ ]  We need to decide whether we want to group subsequent createNode actions by time interval or so, to avoid having too many of those createNode actions spamming the version restore view.

Do you want to do this in this PR? I'd suggest to leave it the current state and iterate in another PR. Ideally, I would also like some grouping by date, so that there are collapsibles for each day. Then we could also remove the date formatting from each version so that there are only the times of the day. So, like that:

```
Monday, 19-07-18
   Version X (15:23)
   Version X (8:53)
Friday, 16-07-18
   Version X (18:23)
   Version X (12:53)
```
